File: src/main/java/it/unibo/crabinv/Controller/save/SessionController.java
package it.unibo.crabinv.Controller.save;

import it.unibo.crabinv.Model.save.GameSession;
import it.unibo.crabinv.Model.save.Save;

public interface SessionController {

    /**
     * Returns the Save used by the SessionController
     * @return the Save used by the SessionController
     */
    Save getSave();

    /**
     * Returns the current GameSession (if it exists)
     * @return the current GameSession, can return null
     */
    GameSession getGameSession();

    /**
     * Checks if there is an active GameSession bound to the save and returns it,
     * if there is no active GameSession creates a new GameSession and binds it to the save
     * @return the loaded or newly created GameSession
     */
    GameSession newGameSession();

    /**
     * Manages the GameSession logic after a GameOver,
     * saves a SessionRecord,
     * inserts in into the PlayerMemorial
     * and deletes the GameSession
     */
    void gameOverGameSession();

}

File: src/main/java/it/unibo/crabinv/controller/core/save/SessionController.kt
package it.unibo.crabinv.controller.core.save

import it.unibo.crabinv.model.core.save.GameSession
import it.unibo.crabinv.model.core.save.Save

/**
 * Contract for a controller to manage Session-related operations.
 */
interface SessionController {
    /**
     * @return the Save used by the SessionController
     */
    val save: Save

    /**
     * @return the current GameSession, can return null
     */
    val gameSession: GameSession

    /**
     * Checks if there is an active GameSession bound to the save and returns it.
     * If there is no active GameSession creates a new GameSession and binds it to the save
     * @return the loaded or newly created GameSession
     */
    fun newGameSession(): GameSession

    /**
     * Manages the GameSession logic after a GameOver.
     * Saves a SessionRecord, inserts into the PlayerMemorial and deletes the GameSession
     */
    fun gameOverGameSession()
}
